hook/scriptwrapper.go

```diff
@@ -20,26 +20,26 @@ const (
 
 	batchScript = `@echo off
 SETLOCAL ENABLEDELAYEDEXPANSION
-SET > "{{.BeforeEnvFileName}}"
+buildkite-agent env > "{{.BeforeEnvFileName}}"
 CALL "{{.PathToHook}}"
 SET BUILDKITE_HOOK_EXIT_STATUS=!ERRORLEVEL!
 SET BUILDKITE_HOOK_WORKING_DIR=%CD%
-SET > "{{.AfterEnvFileName}}"
+buildkite-agent env > "{{.AfterEnvFileName}}"
 EXIT %BUILDKITE_HOOK_EXIT_STATUS%`
 
 	powershellScript = `$ErrorActionPreference = "STOP"
-Get-ChildItem Env: | Foreach-Object {"$($_.Name)=$($_.Value)"} | Set-Content "{{.BeforeEnvFileName}}"
+buildkite-agent env | Set-Content "{{.BeforeEnvFileName}}"
 {{.PathToHook}}
 if ($LASTEXITCODE -eq $null) {$Env:BUILDKITE_HOOK_EXIT_STATUS = 0} else {$Env:BUILDKITE_HOOK_EXIT_STATUS = $LASTEXITCODE}
 $Env:BUILDKITE_HOOK_WORKING_DIR = $PWD | Select-Object -ExpandProperty Path
-Get-ChildItem Env: | Foreach-Object {"$($_.Name)=$($_.Value)"} | Set-Content "{{.AfterEnvFileName}}"
+buildkite-agent env | Set-Content "{{.AfterEnvFileName}}"
 exit $Env:BUILDKITE_HOOK_EXIT_STATUS`
 
-	bashScript = `export -p > "{{.BeforeEnvFileName}}"
+	bashScript = `buildkite-agent env > "{{.BeforeEnvFileName}}"
```

Comment:
Does this introduce a dependency on (the correct / currently-used version of) `buildkite-agent` being in `$PATH`? Or do we already do that elsewhere? Might need to provide an absolute path or something?

I'm wondering what happens if I `go build -o buildkite-agent && buildkite-agent start` on a system that has an older version installed in `$PATH`… I think it'll be surprising if the running version calls the older version, especially if the older version doesn't _have_ the `env` subcommand.

Reply:
So looking at other parts of the bootstrap, we definitely do depend on the agent being on the PATH:
https://github.com/buildkite/agent/blob/main/bootstrap/bootstrap.go#L1866-L1868

but i think you're right to question this, and we should probably stay away from assuming the agent is on the path in the future, it'll only cause trouble when testing things locally.

We should be able to get away with using `os.Executable()` for this with no worries

Reply:
oh weird! i _thought_ we could use `os.Executable()` here, but for some reason it doesn't appear to play nicely with [bintest](https://github.com/buildkite/bintest), and causes every(?) test with bintest to fail with errors like
```
hooks_integration_test.go:450: fork/exec /opt/homebrew/bin/git: resource temporarily unavailable
# or
ssh_test.go:88: Compile of _bintest_06ec6d25081df5c7f7c59c7397831c7eae5da454/main.go failed: go: error obtaining buildID for go tool compile: fork/exec /opt/homebrew/Cellar/go/1.19.2/libexec/pkg/tool/darwin_arm64/compile: resource temporarily unavailable
# or
shell_test.go:265: Error running `/bin/pwd`: fork/exec /bin/pwd: resource temporarily unavailable
```

... and then the test hangs for 10 minutes, then times out, and everything is terrible forever. This only happens if i add an `os.Executable()` call to the scriptwrapper.

I think i'm going to close this can of worms by just calling `buildkite-agent env` directly, as there's prior art there, but i might make a note of it in a comment somewhere?